src/Core/Injector/Injector.php

```diff
@@ -966,12 +971,13 @@ public function unregisterObjects($types)
      *
      * @throws NotFoundExceptionInterface  No entry was found for **this** identifier.
      *
-     * @param string $name The name of the service to retrieve. If not a registered
+     * @template T of object
+     * @param class-string<T> $name The name of the service to retrieve. If not a registered
      * service, then a class of the given name is instantiated
      * @param bool $asSingleton If set to false a new instance will be returned.
      * If true a singleton will be returned unless the spec is type=prototype'
      * @param array $constructorArgs Args to pass in to the constructor. Note: Ignored for singletons
-     * @return mixed Instance of the specified object
+     * @return ($name is class-string<T> ? T : mixed) Instance of the specified object
```

Comment:
Tried this in my IDE - it's worth noting that intelephense doesn't seem to understand these conditional returns, so we may not be able to merge this style into core.

Reply:
Simply `@return class-string<T>` works but of course that won't always be correct. Could arguably be better than nothing though, even if PHPstan doesn't like it.

Reply:
This is where the `@phpstan-return` or `@psalm-return` would come in handy. Only IDEs and phpstan/psalm will understand the prefixed return type while everything else will read the `@return`

```suggestion
     * @return mixed Instance of the specified object
     * @phpstan-return ($name is class-string<T> ? T : mixed) Instance of the specified object
```